### PR TITLE
[cld3] New Port

### PR DESCRIPTION
**Pull request by rozdaybeda**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by LilyWangLL** [comment]
Convert this PR to draft since there are CI pipeline errors need to fix. Please ping us if this PR is ready for review again.

**Comment by LilyWangLL** [comment]
Pinging @rozdaybeda for response. Is work still being done for this PR?

File: scripts/ci.baseline.txt [hunk 1]
```diff
@@ -1391,6 +1391,13 @@ ztd-text:arm-neon-android=fail
 ztd-text:arm64-android=fail
 ztd-text:x64-android=fail
 zyre:x64-windows-static-md=fail
+cld3:arm-neon-android=fail
+cld3:arm64-android=fail
+cld3:x64-android=fail
+cld3:x64-linux=fail
+tesseract:x64-android=fail
+tesseract:arm64-android=fail
+tesseract:arm-neon-android=fail
```
Comment (LilyWangLL) [inline scripts/ci.baseline.txt:1400]:
```suggestion
```
The android building of `tesseract` has been fixed in https://github.com/microsoft/vcpkg/pull/33525.

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

File: ports/cld3/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,22 @@
+vcpkg_from_github(
```
Comment (LilyWangLL) [inline ports/cld3/portfile.cmake:1]:
```suggestion
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)

vcpkg_from_github(
```

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Comment by rozdaybeda** [comment]
> Pinging @rozdaybeda for response. Is work still being done for this PR?
I finished working on PR

**Review by LilyWangLL — DISMISSED** [review]

File: ports/cld3/fix-install.patch [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ add_definitions(-std=c++11) # Needed for std::to_string(), ...
+ 
++if(MSVC_VERSION GREATER_EQUAL "1930")
++set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /std:c++20")
++endif(MSVC)
++
+ include_directories(${CMAKE_CURRENT_BINARY_DIR} ${Protobuf_INCLUDE_DIRS}) # needed to include generated pb headers
+ 
+-add_library(${PROJECT_NAME} 
++add_library(${PROJECT_NAME} STATIC
```
Comment (dg0yt) [inline ports/cld3/fix-install.patch:16]: Is this change necessary, given `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)`?
But ...

File: ports/cld3/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,24 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```
Comment (dg0yt) [inline ports/cld3/portfile.cmake:1]: ... is this generally necessary, or should this be limited to `WIN32`?

File: scripts/ci.baseline.txt [hunk 5]
```diff
@@ -1391,6 +1391,10 @@ ztd-text:arm-neon-android=fail
 ztd-text:arm64-android=fail
 ztd-text:x64-android=fail
 zyre:x64-windows-static-md=fail
+cld3:arm-neon-android=fail
+cld3:arm64-android=fail
+cld3:x64-android=fail
+cld3:x64-linux=fail
```
Comment (dg0yt) [inline scripts/ci.baseline.txt:1397]: Why does it fail? I would prefer to see linux fixed. Unfortunately, there seems to be no CI log for linux because all builds were canceled too early.

Comment (rozdaybeda) [reply scripts/ci.baseline.txt]: @dg0yt linux failed with an error: 
##[error]PowerShell wrote one or more lines to the standard error stream.

File: ports/cld3/fix-install.patch (hunk 3 above)
Comment (rozdaybeda) [reply ports/cld3/fix-install.patch]: I agree. I removed this change: add_library(${PROJECT_NAME} ~STATIC~

File: ports/cld3/portfile.cmake (hunk 4 above)
Comment (rozdaybeda) [reply ports/cld3/portfile.cmake:1]: I think the сld3 library interface is not designed to be used as a shared library.

Comment (dg0yt) [reply ports/cld3/portfile.cmake:1]: I don't see any indicators for this limitation, on non-windows platforms.

File: ports/cld3/fix-install.patch [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+index 2fa3908..de5d149 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -21,12 +21,11 @@ my_protobuf_generate_cpp(cld_3/protos PROTO_SRCS PROTO_HDRS src/feature_extracto
+ message(STATUS "PROTO_HDRS= ${PROTO_HDRS}")
+ 
+ add_definitions(-fPIC) # Position Independant Code
+-add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
+-add_definitions(-std=c++11) # Needed for std::to_string(), ...
++set (CMAKE_CXX_STANDARD 20)
```
Comment (dg0yt) [inline ports/cld3/fix-install.patch:11]: Does setting C++20 here affect requirements for consumers? Isn't C++11 enough?
(The original CMake code was bad, passing compiler options in the definitions...)

File: scripts/ci.baseline.txt (hunk 5 above)
Comment (dg0yt) [reply scripts/ci.baseline.txt]: The failure logs are freely available as "artifacts" on the AZP site of the build. The download is hidden behind three dots when hovering an item:
https://dev.azure.com/vcpkg/public/_build/results?buildId=93990&view=artifacts&pathAsName=false&type=publishedArtifacts

So this is for linux:
~~~
[36/38] : && /usr/bin/c++ -fPIC -g -rdynamic CMakeFiles/language_identifier_features_test.dir/src/language_identifier_features_test.cc.o -o language_identifier_features_test  libcld3.a  /mnt/vcpkg-ci/installed/x64-linux/debug/lib/libprotobuf-lited.a && :
FAILED: language_identifier_features_test 
: && /usr/bin/c++ -fPIC -g -rdynamic CMakeFiles/language_identifier_features_test.dir/src/language_identifier_features_test.cc.o -o language_identifier_features_test  libcld3.a  /mnt/vcpkg-ci/installed/x64-linux/debug/lib/libprotobuf-lited.a && :
/usr/bin/ld: CMakeFiles/language_identifier_features_test.dir/src/language_identifier_features_test.cc.o: in function `google::protobuf::internal::ArenaStringPtr::Set(char const*, google::protobuf::Arena*)':
/mnt/vcpkg-ci/installed/x64-linux/include/google/protobuf/arenastring.h:409: undefined reference to `google::protobuf::internal::ArenaStringPtr::Set(std::string const&, google::protobuf::Arena*)'
/usr/bin/ld: libcld3.a(sentence.pb.cc.o): in function `chrome_lang_id::Sentence::Sentence(google::protobuf::internal::ConstantInitialized)':
/mnt/vcpkg-ci/buildtrees/cld3/x64-linux-dbg/cld_3/protos/sentence.pb.cc:22: undefined reference to `google::protobuf::internal::fixed_address_empty_string'
/usr/bin/ld: /mnt/vcpkg-ci/buildtrees/cld3/x64-linux-dbg/cld_3/protos/sentence.pb.cc:22: undefined reference to `google::protobuf::internal::fixed_address_empty_string'
/usr/bin/ld: libcld3.a(sentence.pb.cc.o): in function `chrome_lang_id::Token::Token(google::protobuf::internal::ConstantInitialized)':
/mnt/vcpkg-ci/buildtrees/cld3/x64-linux-dbg/cld_3/protos/sentence.pb.cc:39: undefined reference to `google::protobuf::internal::fixed_address_empty_string'
/usr/bin/ld: /mnt/vcpkg-ci/buildtrees/cld3/x64-linux-dbg/cld_3/protos/sentence.pb.cc:39: undefined reference to `google::protobuf::internal::fixed_address_empty_string'
/usr/bin/ld: /mnt/vcpkg-ci/buildtrees/cld3/x64-linux-dbg/cld_3/protos/sentence.pb.cc:39: undefined reference to `google::protobuf::internal::fixed_address_empty_string'
/usr/bin/ld: libcld3.a(sentence.pb.cc.o):/mnt/vcpkg-ci/buildtrees/cld3/x64-linux-dbg/cld_3/protos/sentence.pb.cc:39: more undefined references to `google::protobuf::internal::fixed_address_empty_string' follow
~~~

Is linking to protobuf-lite really enough in static linkage?

File: ports/cld3/fix-install.patch (hunk 6 above)
Comment (rozdaybeda) [reply ports/cld3/fix-install.patch]: > Does setting C++20 here affect requirements for consumers?
No
> Isn't C++11 enough?
I got an error with C++11:
C:\work\vcpkg\buildtrees\cld3\src\c4f96dc661-2649dceb5d.clean\src\sentence_features.cc(27): error C2766: explicit specialization; 'registry_' has already been defined

I returned version 11 and made another fix

Comment (rozdaybeda) [reply ports/cld3/fix-install.patch]: @LilyWangLL, @dg0yt  ping

**Review by LilyWangLL — APPROVED** [review]

File: ports/cld3/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,26 @@
+IF (WIN32)
+  vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+ENDIF()
```
Comment (LilyWangLL) [inline ports/cld3/portfile.cmake:3]:
```suggestion
if (VCPKG_TARGET_IS_WINDOWS)
  vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
endif()
```

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

File: ports/cld3/vcpkg.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  "license": "Apache-2.0",
+  "dependencies": [
+    "protobuf",
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
```
Comment (LilyWangLL) [inline ports/cld3/vcpkg.json:15]:
```suggestion
```

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Comment by LilyWangLL** [comment]
https://github.com/google/cld3/blob/b48dc46512566f5a2d41118c8c1116c4f96dc661/CMakeLists.txt#L61-L69

Are the following tools useful?
image

**Comment by rozdaybeda** [comment]
> https://github.com/google/cld3/blob/b48dc46512566f5a2d41118c8c1116c4f96dc661/CMakeLists.txt#L61-L69
> 
> Are the following tools useful? image

Not really, remove them from the build?

**Comment by LilyWangLL** [comment]
> Not really, remove them from the build?

These tools are installed to the temp folder, I just want confirm do these tools need to installed to packages and installed folder.

**Comment by rozdaybeda** [comment]
> > Not really, remove them from the build?
> 
> These tools are installed to the temp folder, I just want confirm do these tools need to installed to packages and installed folder.

I don't think they are needed

File: ports/cld3/vcpkg.json (hunk 8 above)
Comment (LilyWangLL) [reply ports/cld3/vcpkg.json:15]: Please fix this, the `portfile.cmake` of this port doesn't use `vcpkg_cmake_config_fixup`.

Comment (rozdaybeda) [reply ports/cld3/vcpkg.json]: done

**Review by LilyWangLL — DISMISSED** [review]

**Comment by BillyONeal** [comment]
I would like other maintainers opinions on the checked in file lists

File: ports/cld3/fix-install.patch [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+@@ -67,3 +67,65 @@ target_link_libraries(getonescriptspan_test cld3 ${Protobuf_LITE_LIBRARIES})
+ 
+ add_executable(language_identifier_features_test src/language_identifier_features_test.cc)
+ target_link_libraries(language_identifier_features_test cld3 ${Protobuf_LITE_LIBRARIES})
++
++################################################################################
++# Dependencies
++################################################################"################
++
++set(PUBLIC_HEADER
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply ports/cld3/fix-install.patch:23]: OK I spoke with @ras0219-msft about this one and we think this is OK.

**Review by BillyONeal — DISMISSED** [review]
Can you explain what fix-install.patch is doing?

**Comment by rozdaybeda** [comment]
> Can you explain what fix-install.patch is doing?

two things: 
fixes the build for Linux 
adds an installation section to cmake

File: scripts/ci.baseline.txt [hunk 10]
```diff
@@ -1437,4 +1440,4 @@ vcpkg-ci-wxwidgets:x64-osx=pass
 vcpkg-ci-wxwidgets:x64-windows-static-md=pass
 vcpkg-ci-wxwidgets:x64-windows-static=pass
 vcpkg-ci-wxwidgets:x64-windows=pass
-vcpkg-ci-wxwidgets:x86-windows=pass
```
Comment (data-queue) [inline scripts/ci.baseline.txt:1440]: Can we remove the line ending change in this file?

File: ports/cld3/fix-install.patch [hunk 11]
```diff
@@ -0,0 +1,92 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 2fa3908..191adf8 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -21,7 +21,7 @@ my_protobuf_generate_cpp(cld_3/protos PROTO_SRCS PROTO_HDRS src/feature_extracto
+ message(STATUS "PROTO_HDRS= ${PROTO_HDRS}")
+ 
+ add_definitions(-fPIC) # Position Independant Code
+-add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
```
[the remark this answers is not in the document]
Comment (data-queue) [reply ports/cld3/fix-install.patch]: @BillyONeal It's a dual ABI macro. I think we should just remove the line, since the default value of the macro is `1`.

**Review by LilyWangLL — DISMISSED** [review]

File: scripts/ci.baseline.txt [hunk 12]
```diff
@@ -1298,6 +1298,9 @@ ztd-text:arm-neon-android=fail
 ztd-text:arm64-android=fail
 ztd-text:x64-android=fail
 zyre:x64-windows-static-md=fail
+cld3:arm-neon-android=fail
```
Comment (BillyONeal) [inline scripts/ci.baseline.txt:1301]: Stupid question: What in our android test setup is the problem?

File: ports/cld3/fix-build.patch [hunk 13]
```diff
@@ -0,0 +1,33 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 2fa3908..ca94bcb 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -21,7 +21,6 @@ my_protobuf_generate_cpp(cld_3/protos PROTO_SRCS PROTO_HDRS src/feature_extracto
+ message(STATUS "PROTO_HDRS= ${PROTO_HDRS}")
+ 
+ add_definitions(-fPIC) # Position Independant Code
+-add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
```
Comment (JavierMatosD) [inline ports/cld3/fix-build.patch:9]: Is there a reason to remove this?

**Comment by data-queue** [comment]
I'm not sure why the `license/cla` check is still waiting, when it was passing before

**Comment by dg0yt** [comment]
> I'm not sure why the `license/cla` check is still waiting, when it was passing before

It is not AZP but GHA. Close+reopen might re-trigger it.

File: ports/cld3/portfile.cmake [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
+  vcpkg_cmake_get_vars(cmake_vars_file)
+  include("${cmake_vars_file}")
+  if (VCPKG_DETECTED_CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+    set(VCPKG_C_FLAGS "${VCPKG_C_FLAGS} -DCOMPILER_MSVC")
+    set(VCPKG_CXX_FLAGS "${VCPKG_CXX_FLAGS} -DCOMPILER_MSVC")
+  endif()
+endif()
```
Comment (dg0yt) [inline ports/cld3/portfile.cmake:22]: This can be done in the build, via patch or `CMAKE_PROJECT_INCLUDE`. Saves build time.

File: ports/cld3/portfile.cmake [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+  SOURCE_PATH "${SOURCE_PATH}"
+  OPTIONS ${FEATURE_OPTIONS}
+    -DINSTALL_LIB_DIR:STRING=lib
+)
+
+vcpkg_cmake_install()
+vcpkg_fixup_pkgconfig()
+vcpkg_copy_pdbs()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
```
Comment (dg0yt) [inline ports/cld3/portfile.cmake:34]: Probably not needed when it doesn't install headers.

File: ports/cld3/fix-build.patch [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
++#add_executable(getonescriptspan_test src/script_span/getonescriptspan_test.cc)
++#target_link_libraries(getonescriptspan_test cld3 ${Protobuf_LITE_LIBRARIES})
+ 
+-add_executable(language_identifier_features_test src/language_identifier_features_test.cc)
+-target_link_libraries(language_identifier_features_test cld3 ${Protobuf_LITE_LIBRARIES})
++#add_executable(language_identifier_features_test src/language_identifier_features_test.cc)
++#target_link_libraries(language_identifier_features_test cld3 ${Protobuf_LITE_LIBRARIES})
++
++install(TARGETS ${PROJECT_NAME} LIBRARY DESTINATION "${INSTALL_LIB_DIR}" )
+\ No newline at end of file
```
Comment (dg0yt) [inline ports/cld3/fix-build.patch:34]: Shortcut:
```suggestion
@@ -59,5 +59,7 @@ add_library(${PROJECT_NAME}
 	)
 
+install(TARGETS ${PROJECT_NAME} LIBRARY DESTINATION "${INSTALL_LIB_DIR}" )
+return()
+
 # unit tests exec:
 add_executable(language_identifier_main src/language_identifier_main.cc)
 target_link_libraries(language_identifier_main cld3 ${Protobuf_LITE_LIBRARIES})
```

File: ports/cld3/portfile.cmake [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+file(GLOB PUBLIC_HEADER_FILES LIST_DIRECTORIES false "${SOURCE_PATH}/src/*.h")
+file(INSTALL ${PUBLIC_HEADER_FILES} DESTINATION "${CURRENT_PACKAGES_DIR}/include/cld3")
+
+file(GLOB HEADERS_SCRIPT_SPAN_FILES LIST_DIRECTORIES false "${SOURCE_PATH}/src/script_span/*.h")
+file(INSTALL ${HEADERS_SCRIPT_SPAN_FILES} DESTINATION "${CURRENT_PACKAGES_DIR}/include/cld3/script_span")
+
+file(GLOB HEADERS_PROTO_FILES LIST_DIRECTORIES false "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/cld_3/protos/*.h")
+file(INSTALL ${HEADERS_PROTO_FILES} DESTINATION "${CURRENT_PACKAGES_DIR}/include/cld_3/protos")
+
+vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```
Comment (dg0yt) [inline ports/cld3/portfile.cmake:45]: Please add another linebreak.

**Review by dg0yt — CHANGES_REQUESTED** [review]

File: ports/cld3/fix-build.patch (hunk 13 above)
Comment (LilyWangLL) [reply ports/cld3/fix-build.patch:9]: Pinging @rozdaybeda for response. Could you please explain the reason of this behavior?

**Comment by BillyONeal** [comment]
Sorry I broke the build lab Sunday night

File: ports/cld3/fix-build.patch (hunk 13 above)
Comment (rozdaybeda) [reply ports/cld3/fix-build.patch]: @LilyWangLL  sorry it was this case https://github.com/orgs/community/discussions/10369 )

**Comment by dg0yt** [comment]
@JavierMatosD This new port doesn't meet the guidelines: It adds CMake config which is not from upstream (AFAICS), but which lacks the `unofficial` prefix. I thought this would be checked by the official reviewers? This should be fixed immediately.

**Comment by JavierMatosD** [comment]
> @JavierMatosD This new port doesn't meet the guidelines: It adds CMake config which is not from upstream (AFAICS), but which lacks the `unofficial` prefix. I thought this would be checked by the official reviewers? This should be fixed immediately.

You are correct. I missed that. I'll open a PR with the fix shortly. Thanks for the heads up!
